### PR TITLE
Add pagination metadata to the API

### PR DESCRIPTION
**Pull request by ryzngard**
Create a new /pagedreplays  api to get replays with pagination information. The structure returned is outlined below. It shares the same code as /replays, with the addition of taking in a $page parameter to indicate what page should be retrieved. 

{ 
    "page": number,
    "page_count": number,
    "total": number,
    "per_page": number
   "Replays": []
}

### CODE REVIEW

**Comment by ryzngard** [comment]
This hasn't been tested yet, but is approximately what changes need to be made to support this. I'm going through the documentation for testing now to verify before this can be merged.

File: app/Http/Controllers/ReplayController.php [hunk 1]
```diff
@@ -45,45 +49,26 @@ public function store(Request $request, ReplayService $replayService)
...
+     * Show replay list with page metadata
+     * 
+     * @param Request $request
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function paged(Request $request)
+    {
+        $total = $query->count();
+        $pageCount = $total / ReplayController::PAGE_SIZE;
+        $metadata = ['per_page' => ReplayController::PAGE_SIZE, 'page' => $page, 'page_count' => $pageCount, 'total' => $pageCount];
```
Comment (poma) [inline app/Http/Controllers/ReplayController.php]: Shouldn't it be 'total' => $total?

**Comment by poma** [comment]
I suggest moving it to `replays/paged` and placing metadata in a root element (remove `_metadata`)

Also move `->orderBy('id')` to `getQuery()` since it is called in both functions.

File: app/Http/Controllers/ReplayController.php [hunk 2]
```diff
@@ -63,10 +63,8 @@ public function paged(Request $request)
...
         $total = $query->count();
         $pageCount = $total / ReplayController::PAGE_SIZE;
-        $metadata = ['per_page' => ReplayController::PAGE_SIZE, 'page' => $page, 'page_count' => $pageCount, 'total' => $pageCount];
-        $replays = $query->orderBy('id')->forPage($page, ReplayController::PAGE_SIZE)->get();
-
-        $response = ['_metadata' => $metadata, 'Replays' => $replays];
+        $replays = $query->forPage($page, ReplayController::PAGE_SIZE)->get();
+        $response = ['per_page' => ReplayController::PAGE_SIZE, 'page' => $page, 'page_count' => $pageCount, 'total' => $total, 'Replays' => $replays];
 
         return $response()->json();
```
Comment (poma) [inline app/Http/Controllers/ReplayController.php]: Does this work? Shouldn't it be `response()->json($response);`?

**Comment by poma** [comment]
I see a couple of syntax errors here. Please check locally that this code runs without exceptions.

**Comment by ryzngard** [comment]
I haven't been able to get this running locally. I've never done php before, so bare with me if this is obvious. I tried getting running with WSL on Windows (Ubuntu) using the chef options. Highlight of the error and full dump below. I'll try running in a VM with a full Ubuntu installation once I find the time. Is there a way to just compile the PHP to catch errors? I _think_ you can, but that's not a normal use case. A little research shows that it's possible, but I don't know how feasible. 

> ---- Begin output of ./deploy.sh ----
> STDOUT: Application is now in maintenance mode.
> Already up-to-date.
> 
> 
>   [Illuminate\Database\QueryException]
>   SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from inform
>   ation_schema.tables where table_schema = hotsapi and table_name = migration
>   s)
> 
> 
> 
>   [Doctrine\DBAL\Driver\PDOException]
>   SQLSTATE[HY000] [2002] No such file or directory
> 
> 
> 
>   [PDOException]
>   SQLSTATE[HY000] [2002] No such file or directory
> STDERR: From https://github.com/poma/hotsapi
>  * branch            master     -> FETCH_HEAD
> Loading composer repositories with package information
> Installing dependencies (including require-dev) from lock file
> Nothing to install or update
> Generating optimized autoload files
>  Illuminate\Foundation\ComposerScripts::postAutoloadDump
> ---- End output of ./deploy.sh ----
> Ran ./deploy.sh returned 1
> [2017-09-16T08:39:18+03:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
> [2017-09-16T08:39:18+03:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)


Full output of bootstrap.sh

> /mnt/c/Users/ryzng/Source/Repos/hotsapi.chef$ sudo ./bootstrap.sh
> Reading package lists... Done
> Building dependency tree
> Reading state information... Done
> curl is already the newest version (7.47.0-1ubuntu2.2).
> git is already the newest version (1:2.7.4-0ubuntu1.2).
> 0 upgraded, 0 newly installed, 0 to remove and 54 not upgraded.
> Resolving cookbook dependencies...
> Fetching 'hotsapi' from source at .
> Using apt (6.1.4)
> Using hotsapi (0.1.1) from source at .
> Using ssh_authorized_keys (0.4.0)
> Using locale (2.0.1)
> Using usability (1.0.0) from https://github.com/poma/chef-usability.git (at master)
> Vendoring apt (6.1.4) to ./cookbooks/apt
> Vendoring hotsapi (0.1.1) to ./cookbooks/hotsapi
> Vendoring locale (2.0.1) to ./cookbooks/locale
> Vendoring ssh_authorized_keys (0.4.0) to ./cookbooks/ssh_authorized_keys
> Vendoring usability (1.0.0) to ./cookbooks/usability
> [2017-09-16T08:37:11+03:00] WARN: No config file found or specified on command line, using command line options.
> Starting Chef Client, version 13.2.20
> [2017-09-16T08:37:52+03:00] WARN: Plugin Network: unable to detect ipaddress
> [2017-09-16T08:37:52+03:00] WARN: Plugin Network: unable to detect ipaddress
> resolving cookbooks for run list: ["usability", "usability::zsh", "usability::ssh-key", "hotsapi"]
> Synchronizing Cookbooks:
>   - usability (1.0.0)
>   - locale (2.0.1)
>   - apt (6.1.4)
>   - hotsapi (0.1.1)
>   - ssh_authorized_keys (0.4.0)
> Installing Cookbook Gems:
> Compiling Cookbooks...
> Converging 61 resources
> Recipe: locale::default
>   * execute[Generate locale] action run (skipped due to not_if)
>   * execute[Update locale] action run (skipped due to not_if)
> Recipe: apt::default
>   * file[/var/lib/apt/periodic/update-success-stamp] action nothing (skipped due to action :nothing)
>   * apt_update[periodic] action periodic (up to date)
>   * execute[apt-get update] action nothing (skipped due to action :nothing)
>   * execute[apt-get autoremove] action nothing (skipped due to action :nothing)
>   * execute[apt-get autoclean] action nothing (skipped due to action :nothing)
>   * directory[/var/cache/local] action create (up to date)
>   * directory[/var/cache/local/preseeding] action create (up to date)
>   * template[/etc/apt/apt.conf.d/10dpkg-options] action create (up to date)
>   * template[/etc/apt/apt.conf.d/10recommends] action create (up to date)
>   * apt_package[apt-transport-https] action install (up to date)
> Recipe: usability::default
>   * apt_package[ne] action install (up to date)
>   * apt_package[tree] action install (up to date)
>   * apt_package[git] action install (up to date)
>   * apt_package[screen] action install (up to date)
>   * apt_package[curl] action install (up to date)
>   * apt_package[man-db] action install (up to date)
>   * apt_package[telnet] action install (up to date)
>   * apt_package[lsof] action install (up to date)
>   * apt_package[zip] action install (up to date)
>   * apt_package[unzip] action install (up to date)
>   * apt_package[bash-completion] action install (up to date)
>   * apt_package[traceroute] action install (up to date)
>   * apt_package[htop] action install (up to date)
>   * link[/etc/localtime] action create (up to date)
>   * cookbook_file[/etc/timezone] action create (up to date)
>   * apt_package[tzdata] action upgrade (up to date)
>   * cookbook_file[/usr/share/ne/syntax/xml.jsf] action create (up to date)
>   * file[/etc/update-motd.d/10-help-text] action delete (up to date)
>   * file[/etc/update-motd.d/51-cloudguest] action delete (up to date)
> Recipe: usability::zsh
>   * apt_package[zsh] action install (up to date)
>   * apt_package[git] action install (up to date)
>   * git[/opt/oh-my-zsh] action sync (up to date)
>   * git[/opt/oh-my-zsh/custom/plugins/zsh-autosuggestions] action sync (up to date)
>   * git[/opt/oh-my-zsh/custom/plugins/zsh-syntax-highlighting] action sync (up to date)
>   * git[/opt/oh-my-zsh/custom/plugins/zsh-completions] action sync (up to date)
>   * template[/root/.zshrc] action create (up to date)
>   * cookbook_file[/root/.dircolors] action create (up to date)
>   * bash[Set vagrant's shell to zsh] action run
>     - execute "bash"  "/tmp/chef-script20170916-71-111ncjy"
> Recipe: usability::ssh-key
>   * directory[/root/.ssh] action create (up to date)
>   * template[/root/.ssh/authorized_keys] action create (up to date)
> Recipe: hotsapi::default
>   * apt_package[php7.0, php7.0-fpm, php7.0-mysql, php7.0-zip, php7.0-gd, mcrypt, php7.0-mcrypt, php7.0-mbstring, php7.0-xml, php7.0-curl, php7.0-json, composer, zip, unzip, nginx, memcached, php-memcached, supervisor, git] action install (up to date)
>   * service[nginx] action nothing (skipped due to action :nothing)
>   * service[php7.0-fpm] action nothing (skipped due to action :nothing)
>   * execute[hotsapi-git] action run (up to date)
>   * cookbook_file[/etc/nginx/sites-available/default] action create (up to date)
>   * cookbook_file[/etc/nginx/nginx.conf] action create (up to date)
>   * cookbook_file[/etc/php/7.0/fpm/pool.d/www.conf] action create (up to date)
>   * cookbook_file[/etc/php/7.0/fpm/php.ini] action create (up to date)
>   * cookbook_file[/etc/cron.d/laravel] action create (up to date)
>   * template[/var/www/hotsapi/.env] action create
>     - update content in file /var/www/hotsapi/.env from fee5b2 to 9b8475
>     --- /var/www/hotsapi/.env   2017-09-10 22:11:27.484366200 +0300
>     +++ /var/www/hotsapi/.chef-.env20170916-71-19g3014.env      2017-09-16 08:38:10.621811100 +0300
>     @@ -3,8 +3,8 @@
>      APP_ENV=local
>      APP_DEBUG=true
>      APP_LOG_LEVEL=debug
>     -AWS_KEY=
>     -AWS_SECRET=
>     +AWS_KEY={redacted}     
>     +AWS_SECRET={redacted}
>      AWS_REGION=eu-west-1
>      AWS_BUCKET=hotsapi-test
>      DB_HOST=localhost
>   * execute[hotsapi-deploy-script] action run
>     [execute] Application is now in maintenance mode.
>               From https://github.com/poma/hotsapi
>                * branch            master     -> FETCH_HEAD
>               Already up-to-date.
>               Loading composer repositories with package information
>               Installing dependencies (including require-dev) from lock file
>               Nothing to install or update
>               Generating optimized autoload files
>               > Illuminate\Foundation\ComposerScripts::postAutoloadDump
> 
> 
>                 [Illuminate\Database\QueryException]
>                 SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from inform
>                 ation_schema.tables where table_schema = hotsapi and table_name = migration
>                 s)
> 
> 
> 
>                 [Doctrine\DBAL\Driver\PDOException]
>                 SQLSTATE[HY000] [2002] No such file or directory
> 
> 
> 
>                 [PDOException]
>                 SQLSTATE[HY000] [2002] No such file or directory
> 
> 
> 
>     ================================================================================
>     Error executing action `run` on resource 'execute[hotsapi-deploy-script]'
>     ================================================================================
> 
>     Mixlib::ShellOut::ShellCommandFailed
>     ------------------------------------
>     Expected process to exit with [0], but received '1'
>     ---- Begin output of ./deploy.sh ----
>     STDOUT: Application is now in maintenance mode.
>     Already up-to-date.
> 
> 
>       [Illuminate\Database\QueryException]
>       SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from inform
>       ation_schema.tables where table_schema = hotsapi and table_name = migration
>       s)
> 
> 
> 
>       [Doctrine\DBAL\Driver\PDOException]
>       SQLSTATE[HY000] [2002] No such file or directory
> 
> 
> 
>       [PDOException]
>       SQLSTATE[HY000] [2002] No such file or directory
>     STDERR: From https://github.com/poma/hotsapi
>      * branch            master     -> FETCH_HEAD
>     Loading composer repositories with package information
>     Installing dependencies (including require-dev) from lock file
>     Nothing to install or update
>     Generating optimized autoload files
>     > Illuminate\Foundation\ComposerScripts::postAutoloadDump
>     ---- End output of ./deploy.sh ----
>     Ran ./deploy.sh returned 1
> 
>     Resource Declaration:
>     ---------------------
>     # In /home/ryzngard/.chef/local-mode-cache/cache/cookbooks/hotsapi/recipes/default.rb
> 
>      44: execute "hotsapi-deploy-script" do
>      45:   cwd "/var/www/hotsapi"
>      46:   command "./deploy.sh"
>      47:   live_stream true
>      48: end
>      49:
> 
>     Compiled Resource:
>     ------------------
>     # Declared in /home/ryzngard/.chef/local-mode-cache/cache/cookbooks/hotsapi/recipes/default.rb:44:in `from_file'
> 
>     execute("hotsapi-deploy-script") do
>       action [:run]
>       default_guard_interpreter :execute
>       command "./deploy.sh"
>       backup 5
>       cwd "/var/www/hotsapi"
>       returns 0
>       user nil
>       live_stream true
>       declared_type :execute
>       cookbook_name "hotsapi"
>       recipe_name "default"
>       domain nil
>     end
> 
>     System Info:
>     ------------
>     chef_version=13.2.20
>     platform=ubuntu
>     platform_version=16.04
>     ruby=ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]
>     program_name=chef-client worker: ppid=66;start=08:37:16;
>     executable=/opt/chefdk/bin/chef-client
> 
> 
> Running handlers:
> [2017-09-16T08:39:18+03:00] ERROR: Running exception handlers
> [2017-09-16T08:39:18+03:00] ERROR: Running exception handlers
> Running handlers complete
> [2017-09-16T08:39:18+03:00] ERROR: Exception handlers complete
> [2017-09-16T08:39:18+03:00] ERROR: Exception handlers complete
> Chef Client failed. 2 resources updated in 02 minutes 01 seconds
> [2017-09-16T08:39:18+03:00] FATAL: Stacktrace dumped to /home/ryzngard/.chef/local-mode-cache/cache/chef-stacktrace.out
> [2017-09-16T08:39:18+03:00] FATAL: Stacktrace dumped to /home/ryzngard/.chef/local-mode-cache/cache/chef-stacktrace.out
> [2017-09-16T08:39:18+03:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
> [2017-09-16T08:39:18+03:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
> [2017-09-16T08:39:18+03:00] ERROR: execute[hotsapi-deploy-script] (hotsapi::default line 44) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
> ---- Begin output of ./deploy.sh ----
> STDOUT: Application is now in maintenance mode.
> Already up-to-date.
> 
> 
>   [Illuminate\Database\QueryException]
>   SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from inform
>   ation_schema.tables where table_schema = hotsapi and table_name = migration
>   s)
> 
> 
> 
>   [Doctrine\DBAL\Driver\PDOException]
>   SQLSTATE[HY000] [2002] No such file or directory
> 
> 
> 
>   [PDOException]
>   SQLSTATE[HY000] [2002] No such file or directory
> STDERR: From https://github.com/poma/hotsapi
>  * branch            master     -> FETCH_HEAD
> Loading composer repositories with package information
> Installing dependencies (including require-dev) from lock file
> Nothing to install or update
> Generating optimized autoload files
> > Illuminate\Foundation\ComposerScripts::postAutoloadDump
> ---- End output of ./deploy.sh ----
> Ran ./deploy.sh returned 1
> [2017-09-16T08:39:18+03:00] ERROR: execute[hotsapi-deploy-script] (hotsapi::default line 44) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
> ---- Begin output of ./deploy.sh ----
> STDOUT: Application is now in maintenance mode.
> Already up-to-date.
> 
> 
>   [Illuminate\Database\QueryException]
>   SQLSTATE[HY000] [2002] No such file or directory (SQL: select * from inform
>   ation_schema.tables where table_schema = hotsapi and table_name = migration
>   s)
> 
> 
> 
>   [Doctrine\DBAL\Driver\PDOException]
>   SQLSTATE[HY000] [2002] No such file or directory
> 
> 
> 
>   [PDOException]
>   SQLSTATE[HY000] [2002] No such file or directory
> STDERR: From https://github.com/poma/hotsapi
>  * branch            master     -> FETCH_HEAD
> Loading composer repositories with package information
> Installing dependencies (including require-dev) from lock file
> Nothing to install or update
> Generating optimized autoload files
> > Illuminate\Foundation\ComposerScripts::postAutoloadDump
> ---- End output of ./deploy.sh ----
> Ran ./deploy.sh returned 1
> [2017-09-16T08:39:18+03:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
> [2017-09-16T08:39:18+03:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)

**Comment by poma** [comment]
Looks like you didn't configure mysql db. Chef doesn't do it because DB_is external to the web server. Tro to install `mysql-server` first and provide credentials in `chef.json`. Probably I should include an option to install and configure a testing mysql server in chef recipe.

**Comment by poma** [comment]
By the way if you use WSL you'll probably have issues with listening on port 80. If that's the case change port in `/etc/nginx/sites-available/default`

**Comment by poma** [comment]
I've updated hotsapi.chef to configure mysql-server and create default DB. It also automatically generates new APP_KEY now.

**Comment by ryzngard** [comment]
Thanks for all the help! I have at least the site up and running, after a few hours of figuring out how vagrant/homestead work. As expected, there are errors, but at least I now have the environment set up and can look into them. I'm out for work for a few days, but I'll update the PR when I get back and fix the issues

**Comment by poma** [comment]
Manually fixed errors and merged in dfabdef8ef1ea57d97970945e138753f93b48a18

**Pull request closed without merging**